File: skills/agent-swarm/scripts/backends/claude_cli.py
"""Claude CLI Execution Backend; the only module allowed to invoke ``claude``."""

import json
import os
import pathlib
import re
import subprocess

from .base import AgentBackend, ObserveResult, SpawnRequest, SpawnResult, StopRequest


JOB_RE = re.compile(r"backgrounded\s*[·:]\s*(\S+)", re.IGNORECASE)
ATTACH_RE = re.compile(r"claude\s+attach\s+(\S+)")
TERMINAL_STATES = {"done", "completed", "exited", "failed", "stopped", "error", "cancelled"}


def claude_bin():
    override = os.environ.get("AGENT_SWARM_CLAUDE_BIN", "").strip()
    if override:
        return override
    for entry in os.environ.get("PATH", "").split(os.pathsep):
        if not entry:
            continue
        candidate = pathlib.Path(entry) / "claude"
        if not candidate.is_file() or not os.access(str(candidate), os.X_OK):
            continue
        if ".superconductor" in candidate.parts:
            continue
        return str(candidate)
    return "claude"


def _timeout(name, default):
    raw = os.environ.get(name, "").strip()
    try:
        return max(1.0, float(raw)) if raw else default
    except ValueError:
        return default


def _job_id(output):
    for expression in (JOB_RE, ATTACH_RE):
        match = expression.search(output or "")
        if match:
            return match.group(1)
    return None


def _text(value):
    if value is None:
        return ""
    if isinstance(value, bytes):
        return value.decode("utf-8", errors="replace")
    return str(value)


class ClaudeCliBackend(AgentBackend):
    backend_id = "claude_cli"

    def __init__(self, config=None):
        self.config = dict(config or {})

    def _command(self):
        return self.config.get("command") or claude_bin()

    def spawn(self, request):
        if not isinstance(request, SpawnRequest):
            raise TypeError("spawn requires SpawnRequest")
        command = [
            self._command(),
            "--bg",
            "--name",
            request.session_name,
            "--permission-mode",
            "bypassPermissions",
        ]
        if request.model:
            command.extend(["--model", request.model])
        command.append(request.prompt)
        child_env = os.environ.copy()
        child_env.update(request.env or {})
        timeout_seconds = _timeout("AGENT_SWARM_BG_LAUNCH_TIMEOUT_SECONDS", 90.0)
        try:
            completed = subprocess.run(
                command,
                cwd=request.cwd,
                env=child_env,
                stdin=subprocess.DEVNULL,
                capture_output=True,
                text=True,
                check=False,
                timeout=timeout_seconds,
            )
        except subprocess.TimeoutExpired as exc:
            partial = "%s\n%s" % (
                _text(getattr(exc, "stdout", None) or getattr(exc, "output", None)),
                _text(getattr(exc, "stderr", None)),
            )
            job_id = _job_id(partial)
            if job_id:
                return SpawnResult(job_id=job_id, session_name=request.session_name)
            raise RuntimeError(
                "claude --bg timed out after %.1fs without a job id: %s"
                % (timeout_seconds, partial.strip() or "no partial output")
            ) from exc
        output = "%s\n%s" % (completed.stdout or "", completed.stderr or "")
        job_id = _job_id(output)
        if not job_id:
            raise RuntimeError(
                "claude --bg failed (exit=%s): %s"
                % (completed.returncode, output.strip() or "no job id")
            )
        return SpawnResult(job_id=job_id, session_name=request.session_name)

    def stop(self, request):
        if not isinstance(request, StopRequest):
            raise TypeError("stop requires StopRequest")
        job_id = request.job_id
        if not job_id and request.session_name:
            matching = []
            for session in self.list_sessions(cwd=request.cwd):
                name = session.get("name") or session.get("session_name")
                state = session.get("state") or session.get("status")
                if name == request.session_name and state not in TERMINAL_STATES:
                    matching.append(session)
            if len(matching) > 1:
                raise RuntimeError("multiple live Claude sessions match %s" % request.session_name)
            if matching:
                job_id = matching[0].get("job_id") or matching[0].get("id")
                if not job_id:
                    raise RuntimeError("live Claude session %s has no stoppable job id" % request.session_name)
        if not job_id:
            return {"stopped": True, "not_required": True}
        completed = subprocess.run(
            [self._command(), "stop", job_id],
            stdin=subprocess.DEVNULL,
            capture_output=True,
            text=True,
            check=False,
            timeout=_timeout("AGENT_SWARM_AGENT_CONTROL_TIMEOUT_SECONDS", 10.0),
        )
        if completed.returncode != 0:
            raise RuntimeError((completed.stderr or completed.stdout or "claude stop failed").strip())
        return {"stopped": True}

    def observe(self, *, job_id=None, session_name=None, cwd=None):
        if not job_id and not session_name:
            return ObserveResult("unknown", error="session identity is missing")
        try:
            sessions = self.list_sessions(cwd=cwd)
        except Exception as exc:
            return ObserveResult("unknown", error=str(exc))
        matching = []
        for session in sessions:
            if not isinstance(session, dict):
                continue
            candidate_id = session.get("job_id") or session.get("id")
            candidate_name = session.get("name") or session.get("session_name")
            if (job_id and candidate_id == job_id) or (session_name and candidate_name == session_name):
                matching.append(session)
        if not matching:
            return ObserveResult("absent")
        for session in matching:
            value = session.get("state") or session.get("status")
            if value not in TERMINAL_STATES:
                return ObserveResult("present", session=session)
        return ObserveResult("absent", session=matching[-1])

    def list_sessions(self, *, cwd=None):
        completed = subprocess.run(
            [self._command(), "agents", "--json"],
            stdin=subprocess.DEVNULL,
            capture_output=True,
            text=True,
            check=False,
            timeout=_timeout("AGENT_SWARM_AGENT_CONTROL_TIMEOUT_SECONDS", 10.0),
        )
        if completed.returncode != 0:
            raise RuntimeError((completed.stderr or completed.stdout or "claude agents failed").strip())
        try:
            items = json.loads(completed.stdout or "[]")
        except ValueError as exc:
            raise RuntimeError("claude agents returned invalid JSON") from exc
        if not isinstance(items, list):
            raise RuntimeError("claude agents returned a non-array result")
        if cwd:
            expected = os.path.realpath(cwd)
            items = [
                item
                for item in items
                if isinstance(item, dict) and os.path.realpath(item.get("cwd") or "") == expected
            ]
        return items

    def supports_hooks(self):
        return True

File: skills/agents-orchestrator/scripts/backends/claude_cli.ts
import { accessSync, constants as fsConstants, realpathSync, statSync } from "node:fs";
import { delimiter, join } from "node:path";

import { AgentBackend, ObserveResult, SpawnRequest, SpawnResult, StopRequest } from "./base.ts";
import { isRecord, RuntimeError, type RuntimeRecord, ValueError } from "../runtime_types.ts";

const JOB_RE = /backgrounded\s*[·:]\s*(\S+)/iu;
const ATTACH_RE = /claude\s+attach\s+(\S+)/u;
const TERMINAL_STATES = new Set(["done", "completed", "exited", "failed", "stopped", "error", "cancelled"]);

function executable(path: string): boolean {
  try {
    statSync(path);
    accessSync(path, fsConstants.X_OK);
    return true;
  } catch {
    return false;
  }
}

export function claudeBin(environment: Record<string, string | undefined> = process.env): string {
  const override = (environment.AGENTS_ORCHESTRATOR_CLAUDE_BIN ?? "").trim();
  if (override) return override;
  for (const entry of (environment.PATH ?? "").split(delimiter)) {
    if (!entry) continue;
    const candidate = join(entry, "claude");
    if (executable(candidate) && !candidate.split(/[\\/]/u).includes(".superconductor")) return candidate;
  }
  return "claude";
}

function timeout(name: string, fallback: number): number {
  const raw = process.env[name]?.trim();
  const value = raw ? Number(raw) : Number.NaN;
  return Number.isFinite(value) ? Math.max(1, value) : fallback;
}

function jobId(output: string): string | null {
  return JOB_RE.exec(output)?.[1] ?? ATTACH_RE.exec(output)?.[1] ?? null;
}

function run(command: string[], options: { cwd?: string; env?: Record<string, string>; timeout: number }): {
  exitCode: number;
  stdout: string;
  stderr: string;
} {
  const result = Bun.spawnSync(command, {
    cwd: options.cwd,
    env: options.env,
    stdin: "ignore",
    stdout: "pipe",
    stderr: "pipe",
    timeout: options.timeout * 1_000,
  });
  return {
    exitCode: result.exitCode,
    stdout: result.stdout.toString("utf8"),
    stderr: result.stderr.toString("utf8"),
  };
}

export class ClaudeCliBackend extends AgentBackend {
  readonly backendId = "claude_cli";
  readonly config: RuntimeRecord;

  constructor(config: RuntimeRecord = {}) {
    super();
    this.config = { ...config };
  }

  private command(): string {
    return typeof this.config.command === "string" && this.config.command ? this.config.command : claudeBin();
  }

  spawn(request: SpawnRequest): SpawnResult {
    if (!(request instanceof SpawnRequest)) throw new TypeError("spawn requires SpawnRequest");
    const command = [
      this.command(),
      "--bg",
      "--name",
      request.sessionName,
      "--permission-mode",
      "bypassPermissions",
    ];
    if (request.model) command.push("--model", request.model);
    command.push(request.prompt);
    const completed = run(command, {
      cwd: request.cwd,
      env: { ...process.env, ...request.env } as Record<string, string>,
      timeout: timeout("AGENTS_ORCHESTRATOR_BG_LAUNCH_TIMEOUT_SECONDS", 90),
    });
    const identifier = jobId(`${completed.stdout}\n${completed.stderr}`);
    if (!identifier) {
      throw new RuntimeError(
        completed.exitCode === 124
          ? "claude --bg timed out without a job id"
          : `claude --bg failed without a job id (exit=${completed.exitCode})`,
      );
    }
    return new SpawnResult(identifier, request.sessionName);
  }

  stop(request: StopRequest): RuntimeRecord {
    if (!(request instanceof StopRequest)) throw new TypeError("stop requires StopRequest");
    let identifier = request.jobId;
    if (!identifier && request.sessionName) {
      const matching = this.listSessions({ cwd: request.cwd }).filter((session) => {
        const name = session.name ?? session.session_name;
        const state = session.state ?? session.status;
        return name === request.sessionName && !TERMINAL_STATES.has(state);
      });
      if (matching.length > 1) throw new RuntimeError(`multiple live Claude sessions match ${request.sessionName}`);
      if (matching.length === 1) {
        identifier = matching[0]!.job_id ?? matching[0]!.id ?? null;
        if (!identifier) throw new RuntimeError(`live Claude session ${request.sessionName} has no stoppable job id`);
      }
    }
    if (!identifier) return { stopped: true, not_required: true };
    const completed = run([this.command(), "stop", identifier], {
      timeout: timeout("AGENTS_ORCHESTRATOR_AGENT_CONTROL_TIMEOUT_SECONDS", 10),
    });
    if (completed.exitCode !== 0) throw new RuntimeError("claude stop failed");
    return { stopped: true };
  }

  observe(options: { jobId?: string | null; sessionName?: string | null; cwd?: string | null }): ObserveResult {
    if (!options.jobId && !options.sessionName) return new ObserveResult("unknown", null, "session identity is missing");
    let sessions: RuntimeRecord[];
    try {
      sessions = this.listSessions({ cwd: options.cwd });
    } catch {
      return new ObserveResult("unknown", null, "session observation failed");
    }
    const matching = sessions.filter((session) =>
      (options.jobId && (session.job_id ?? session.id) === options.jobId) ||
      (options.sessionName && (session.name ?? session.session_name) === options.sessionName));
    if (matching.length === 0) return new ObserveResult("absent");
    const present = matching.find((session) => !TERMINAL_STATES.has(session.state ?? session.status));
    return present ? new ObserveResult("present", present) : new ObserveResult("absent", matching.at(-1)!);
  }

  listSessions(options: { cwd?: string | null } = {}): RuntimeRecord[] {
    const completed = run([this.command(), "agents", "--json"], {
      timeout: timeout("AGENTS_ORCHESTRATOR_AGENT_CONTROL_TIMEOUT_SECONDS", 10),
    });
    if (completed.exitCode !== 0) throw new RuntimeError("claude agents failed");
    let value: unknown;
    try {
      value = JSON.parse(completed.stdout || "[]") as unknown;
    } catch {
      throw new RuntimeError("claude agents returned invalid JSON");
    }
    if (!Array.isArray(value) || !value.every(isRecord)) throw new RuntimeError("claude agents returned a non-array result");
    if (!options.cwd) return value;
    const expected = realpathSync(options.cwd);
    return value.filter((item) => {
      try {
        return typeof item.cwd === "string" && realpathSync(item.cwd) === expected;
      } catch {
        return false;
      }
    });
  }

  supportsHooks(): boolean {
    return true;
  }
}
